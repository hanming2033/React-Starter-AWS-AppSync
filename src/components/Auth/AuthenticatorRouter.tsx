import { Route, RouteProps, RouteComponentProps, StaticContext, Redirect } from 'react-router'
import React from 'react'
import { AuthProxy } from './AuthProxies/AuthProxy'
import Signin from './SignIn'
import Forgot from './ForgotPassword'
import RequireNewPassword from './RequireNewPassword'
import Signup from './SignUp'
import SignupConfirm from './SignupConfirm'
import { GET_LOCAL_STATES } from '../../data/actions/Queries'
import { GetLocalStatesQuery, SetAuthMutation } from '../../data/graphql-types'
import { ChildProps, graphql, compose } from 'react-apollo'
import VerifyContact from './VerifyContact'
import { IVerification, ICognitoUser } from './AuthProxies/AuthTypes'
import { SET_AUTH } from '../../data/actions/Mutations'
import { JS } from 'aws-amplify'

interface IAuthenticatorProps {
  component:
    | React.ComponentClass<RouteComponentProps<any, StaticContext>>
    | React.StatelessComponent<RouteComponentProps<any, StaticContext>>
    | React.ComponentClass<any>
    | React.StatelessComponent<any>
}

export type validComponents =
  | 'signIn'
  | 'confirmSignIn'
  | 'signUp'
  | 'confirmSignUp'
  | 'forgotPassword'
  | 'requireNewPassword'
  | 'verifyContact'
  | 'TOTPSetup'

export type TtoComp = (newComponent: validComponents, userData?: ICognitoUser) => void
export type TsetAuth = (verification: IVerification) => void

interface IProtectedRouteState {
  userData: ICognitoUser | null
  componentToShow: validComponents
  verification: IVerification
}

export class Authenticator extends React.Component<
  ChildProps<IAuthenticatorProps & RouteProps, GetLocalStatesQuery & SetAuthMutation>,
  IProtectedRouteState
> {
  public state = {
    userData: null,
    componentToShow: 'signUp' as validComponents,
    verification: { verified: {}, unverified: {} }
  }

  public toComp: TtoComp = (component, userData) => {
    this.setState({
      componentToShow: component,
      userData: userData ? userData : this.state.userData
    })
  }

  public setAuth = (verification: IVerification) => {
    const isAuthenticated = !JS.isEmpty(verification.verified)
    this.toComp('verifyContact')
    if (!this.props.mutate) return
    this.props.mutate({ variables: { status: isAuthenticated } })
  }

  public componentDidMount() {
    AuthProxy.checkAuthState()
      .then(res => {
        // console.log('Authenticator Router On Mount: ', res)
        return res
      })
      .then(res => {
        if (!res.data) return
        AuthProxy.verifyUser(res.data.user).then(verification => {
          if (res.data) {
            this.setAuth(verification)
          } else {
            this.setAuth({ verified: {}, unverified: {} })
          }
        })
      })
  }

  public render() {
    const { component: Component, data, ...rest } = this.props
    const { componentToShow } = this.state
    if (!data || !data.auth) return null
    if (data.auth.isAuthenticated && this.props.path === '/authenticate') return <Redirect to="/" />
    if (data.auth.isAuthenticated) return <Route {...rest} render={props => <Component {...props} />} />
    if (componentToShow === 'signIn')
      return <Route {...rest} render={props => <Signin setAuth={this.setAuth} {...props} toComp={this.toComp} />} />
    if (componentToShow === 'signUp') return <Route {...rest} render={props => <Signup {...props} toComp={this.toComp} />} />
    if (componentToShow === 'forgotPassword') return <Route {...rest} render={props => <Forgot {...props} toComp={this.toComp} />} />
    if (componentToShow === 'confirmSignUp') return <Route {...rest} render={props => <SignupConfirm {...props} toComp={this.toComp} />} />
    if (componentToShow === 'requireNewPassword')
      return (
        <Route
          {...rest}
          render={props => <RequireNewPassword setAuth={this.setAuth} {...props} userData={this.state.userData} toComp={this.toComp} />}
        />
      )
    if (componentToShow === 'verifyContact') return <Route {...rest} render={props => <VerifyContact {...props} />} />
    return 'Not Authenticated!' // dummy output
  }
}

export default compose(
  graphql<IAuthenticatorProps & RouteProps, GetLocalStatesQuery>(GET_LOCAL_STATES, { options: { fetchPolicy: 'network-only' } }),
  graphql<IAuthenticatorProps & RouteProps, SetAuthMutation>(SET_AUTH, {})
)(Authenticator)
