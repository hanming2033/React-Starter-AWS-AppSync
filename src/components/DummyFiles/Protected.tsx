import * as React from 'react'
import { AuthProxy } from '../Auth/AuthProxy'
import { RouteComponentProps } from 'react-router'

class Protected extends React.Component<RouteComponentProps<{}>> {
  public render() {
    return (
      <>
        <h3>Private</h3>
        <p>
          Welcome! <button onClick={() => AuthProxy.signOut().then(() => this.props.history.replace('/'))}>Sign Out</button>{' '}
        </p>
      </>
    )
  }
}

export default Protected
