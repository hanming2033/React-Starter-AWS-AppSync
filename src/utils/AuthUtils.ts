import { QueryResult } from 'react-apollo'
import { GetLocalStatesQuery } from '../data/graphql-types'

export const updateCacheForm = (qryRes: QueryResult<GetLocalStatesQuery>, propName: string, value: string) => {
  if (qryRes.data && qryRes.data.forms) {
    const newData: GetLocalStatesQuery = {
      ...qryRes.data,
      forms: {
        ...qryRes.data.forms,
        [propName]: value
      }
    }
    qryRes.client.writeData({ data: newData })
  }
}
