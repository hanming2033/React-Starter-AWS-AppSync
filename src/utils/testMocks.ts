import { GetLocalStatesQuery } from '../data/graphql-types'
import { QueryResult } from 'react-apollo'
import { ApolloError } from 'apollo-client'
import { client } from '../AWSApolloClient'

export const getLocalStateData: GetLocalStatesQuery = {
  auth: {
    __typename: 'auth',
    isAuthenticated: false
  },
  forms: {
    __typename: 'forms',
    input_Email: ''
  },
  nav: {
    __typename: 'nav',
    nextPath: ''
  }
}

export const qryRes: QueryResult<GetLocalStatesQuery> = {
  client,
  data: undefined,
  error: undefined,
  loading: false,
  networkStatus: 7,
  startPolling: jest.fn(),
  stopPolling: jest.fn(),
  subscribeToMore: jest.fn(),
  variables: {},
  refetch: jest.fn(),
  fetchMore: jest.fn(),
  updateQuery: jest.fn()
}

export const apolloError: ApolloError = {
  name: 'error',
  message: 'error',
  graphQLErrors: [{ name: 'error', message: 'error' }],
  networkError: null,
  extraInfo: ''
}
