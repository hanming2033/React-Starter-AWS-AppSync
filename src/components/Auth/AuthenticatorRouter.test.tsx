import * as React from 'react'
import * as enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import enzymeSerializer from 'enzyme-to-json/serializer'
import { getLocalStateData, qryRes } from '../../utils/testMocks'
import { createBrowserHistory, createLocation } from 'history'
import { Authenticator } from './AuthenticatorRouter'
import Protected from '../DummyFiles/Protected'
import { Route } from 'react-router'
import { Query } from '../../../node_modules/react-apollo'

expect.addSnapshotSerializer(enzymeSerializer)
enzyme.configure({ adapter: new Adapter() })

// * always try to test component rendering instead of state unless rendering is happening in child components
// * always try to test on interaction instead of state change unless interaction happens in child components
// * always try to use unit test, snapshot for conditional rendering or maintance cost is high
// * Group by feature. Each feature test on rendering and its interaction
// * -- rendering: how props and state(prefer interaction on component unless interaction happen in children) affect this component
// * -- interaction: how feature's interaction affect itself, other component, state(prefer rendering unless rendering in children), prop method call

const routeProps = {
  history: createBrowserHistory(),
  location: createLocation('/'),
  match: {
    params: '',
    isExact: false,
    path: '',
    url: ''
  },
  staticContext: {
    statusCode: 1
  }
}

describe('<MainComponent /> Main Suite', () => {
  // it('dummy', async () => {
  //   const wrapper = enzyme.shallow(<Authenticator component={Protected} />)
  //   wrapper.setProps({ data: getLocalStateData })
  //   const wrapperInstance = wrapper.instance() as Authenticator
  //   wrapperInstance.toComp('forgotPassword')
  //   const holder = wrapper.find(Route).prop('render')
  //   if (!holder) return
  //   const routeWrapper = enzyme.shallow(holder(routeProps) as React.ReactElement<{}>)
  //   const newRes = { ...qryRes, data: getLocalStateData }
  //   const queryWrapper = enzyme.shallow(routeWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
  //   console.log(queryWrapper.debug())
  // })
  // rendering based on different state and props of main component
  // test here is state and props does not directly affects children
  // test here is state and props affects more than 1 children
  it('snapshot on no data prop', () => {
    const wrapper = enzyme.shallow(<Authenticator component={Protected} />)
    wrapper.setProps({ data: null })
    expect(wrapper).toMatchSnapshot()
  })
  it('snapshot on authenticated & /authenticate redirect', () => {
    const wrapper = enzyme.shallow(<Authenticator component={Protected} />)
    wrapper.setProps({ data: { auth: { isAuthenticated: true } }, path: '/authenticate' })
    expect(wrapper).toMatchSnapshot()
  })
  it('snapshot on authenticated showing actual component', () => {
    const wrapper = enzyme.shallow(<Authenticator component={Protected} />)
    wrapper.setProps({ data: { auth: { isAuthenticated: true } } })
    const holder = wrapper.find(Route).prop('render')
    if (!holder) return
    const routeWrapper = enzyme.shallow(holder(routeProps) as React.ReactElement<{}>)
    expect(routeWrapper).toMatchSnapshot()
  })
  it('snapshot on signUp', () => {
    const wrapper = enzyme.shallow(<Authenticator component={Protected} />)
    wrapper.setProps({ data: { auth: { isAuthenticated: false } } })
    const holder = wrapper.find(Route).prop('render')
    if (!holder) return
    const routeWrapper = enzyme.shallow(holder(routeProps) as React.ReactElement<{}>)
    const newRes = { ...qryRes, data: getLocalStateData }
    const queryWrapper = enzyme.shallow(routeWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
    expect(queryWrapper.childAt(0)).toMatchSnapshot()
  })
  it('snapshot on signIn', () => {
    const wrapper = enzyme.shallow(<Authenticator component={Protected} />)
    wrapper.setProps({ data: getLocalStateData })
    const wrapperInstance = wrapper.instance() as Authenticator
    wrapperInstance.toComp('signIn')
    wrapper.update()
    const holder = wrapper.find(Route).prop('render')
    if (!holder) return
    const routeWrapper = enzyme.shallow(holder(routeProps) as React.ReactElement<{}>)
    const newRes = { ...qryRes, data: getLocalStateData }
    const queryWrapper = enzyme.shallow(routeWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
    expect(queryWrapper.childAt(0)).toMatchSnapshot()
  })
  it('snapshot on forgotPassword', () => {
    const wrapper = enzyme.shallow(<Authenticator component={Protected} />)
    wrapper.setProps({ data: getLocalStateData })
    const wrapperInstance = wrapper.instance() as Authenticator
    wrapperInstance.toComp('forgotPassword')
    wrapper.update()
    const holder = wrapper.find(Route).prop('render')
    if (!holder) return
    const routeWrapper = enzyme.shallow(holder(routeProps) as React.ReactElement<{}>)
    const newRes = { ...qryRes, data: getLocalStateData }
    const queryWrapper = enzyme.shallow(routeWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
    expect(queryWrapper.childAt(0)).toMatchSnapshot()
  })
  it('snapshot on confirmSignUp', () => {})
  it('snapshot on requireNewPassword', () => {})
  it('snapshot on verifyContact', () => {})
  // describe('Feature 1 Suite', () => {
  //   // all things related to this feature/component
  //   describe('Static/Conditional Rendering', () => {
  //     // rendering based on different props and state of this component
  //     it('snapshot on default state/prop', () => {})
  //     it('snapshot on some prop', () => {})
  //     it('snapshot on some state', () => {})
  //   })
  //   describe('Interactions', () => {
  //     // how interaction on this feature affects itself, other component, state(rendering in children), prop method call
  //     it('should call prop method with args xxx', () => {})
  //     it('should cause component x to render', () => {})
  //     it('should cause state to change to xxx', () => {})
  //     // !try to use snapshot for interaction sparingly
  //     it('snapshot on some change', () => {})
  //   })
  // })
})
