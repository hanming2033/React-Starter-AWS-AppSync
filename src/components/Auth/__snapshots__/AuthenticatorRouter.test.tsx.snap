// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainComponent /> Main Suite snapshot on authenticated & /authenticate redirect 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;

exports[`<MainComponent /> Main Suite snapshot on authenticated showing actual component 1`] = `
<React.Fragment>
  <h3>
    Private
  </h3>
  <p>
    Welcome! 
    <button
      onClick={[Function]}
    >
      Sign Out
    </button>
     
  </p>
</React.Fragment>
`;

exports[`<MainComponent /> Main Suite snapshot on no data prop 1`] = `""`;

exports[`<MainComponent /> Main Suite snapshot on signIn 1`] = `
<h1>
  Sign In
</h1>
`;

exports[`<MainComponent /> Main Suite snapshot on signUp 1`] = `
<h1>
  Sign Up
</h1>
`;
